# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [30]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [31]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [32]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [33]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [34]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [35]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [36]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [37]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [38]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [39]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [40]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [41]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [42]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [43]:
# Write your code here
unique_seasons = df['Season'].nunique() #Using nunique function on particular column season to find answer
unique_seasons #Show answer

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [44]:
# Write your code here
most_wins_team = df['Wteam'].value_counts().idxmax() #Use value_counts function on Wteam column with idxmax to find the team with the most wins
most_wins_team #Shows index of the team with the most wins

1181

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [45]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [78]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [47]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [48]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [49]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [50]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [51]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [52]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [53]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [54]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [85]:
# Write your code here
sorted_season = df.sort_values('Season').iloc[100:200][['Season']] #Use sort_values and then iloc to only show index 100-200

sorted_Wteam = df.sort_values('Wteam').iloc[100:200][['Wteam']] #Use sort_values and then iloc to only show index 100-200

sorted_Wscore = df.sort_values('Wscore').iloc[100:200][['Wscore']] #Use sort_values and then iloc to only show index 100-200

#sorted_season #Show sorted season index 100-200
#sorted_Wteam #Show sorted Wteam index 100-200
#sorted_Wscore #Show sorted Wscore index 100-200

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [87]:
# Write your code here
max_win_score = sorted_Wscore['Wscore'].max() #Find highest winning score

max_win_score_row = df[df['Wscore'] == max_win_score] #Find the row the highest winning score is located on

# From the highest winning score row the season and team is found
max_win_score_season = max_win_score_row['Season'].iloc[0]
max_win_score_team = max_win_score_row['Wteam'].iloc[0]

max_win_score_season, max_win_score_team #Show the season and team for the game with the highest winning score

(1985, 1248)

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [57]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [58]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [89]:
# Write your code here
df['ScoreDifference'] = (df['Wscore'] - df['Lscore']).abs() #Find the score difference and create a column
avg_score_diff = df['ScoreDifference'].mean() #Find the average score difference
over_avg_df = df[df['ScoreDifference'] > avg_score_diff] #Filter for games with a greater score difference than the average

over_avg_df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

       HighScoringGame  ScoreDifference  
0                   No               17  
3                   No               16  
6                   No               20  
8                   No               18  
10                 Yes               32  
...                ...              ...  
145280              No               33  
145282              No               22  
145284              No               20  
145285              No               14  
145288              No               13  

[57227 rows x 10 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [91]:
# Write your code here
max_score_diff = over_avg_df['ScoreDifference'].max() #Find max score difference
max_score_diff_row = over_avg_df[over_avg_df['ScoreDifference'] == max_score_diff] #Find row with max score difference

# Extract the season, winning team, and losing team from this row
max_score_diff_season = max_score_diff_row['Season'].iloc[0] #Find season from row
max_score_diff_Wteam = max_score_diff_row['Wteam'].iloc[0] #Find winning team from row
max_score_diff_Lteam = max_score_diff_row['Lteam'].iloc[0] #Find losing team from row

# Print or return the results
max_score_diff_season, max_score_diff_Wteam, max_score_diff_Lteam #show the season and teams involved in the game with the highest ‘ScoreDifference’

(1996, 1409, 1341)

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [61]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [62]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [63]:
df.values

array([[1985, 20, 1228, ..., 64, 'N', 0],
       [1985, 25, 1106, ..., 70, 'H', 0],
       [1985, 25, 1112, ..., 56, 'H', 0],
       ...,
       [2016, 132, 1246, ..., 77, 'N', 1],
       [2016, 132, 1277, ..., 62, 'N', 0],
       [2016, 132, 1386, ..., 74, 'N', 0]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [64]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [94]:
# Write your code here
avg_win_score = df.groupby('Season')['Wscore'].mean() #Grouped by 'Season' column and calculated the mean
avg_win_score #Show average winning score for each season

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [95]:
# Write your code here
max_win_score = df.groupby('Wteam')['Wscore'].max() #Grouped by the 'Wteam' column and calculated max win score for each team
max_win_score #Show maximum winning score for each team across all seasons

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [96]:
# Write your code here
average_scores = df.groupby(['Season', 'Wteam'])['Wscore'].mean() #Grouped by 'Season and 'Wteam' columns and calculated avg win score
average_scores_reset = average_scores.reset_index() #Reset index to sort values
highest_avg_score_season = average_scores_reset.sort_values(['Season', 'Wscore'], ascending=[True, False]) #Sorted by highest average winning score for each season
highest_avg_score_season = highest_avg_score_season.drop_duplicates(subset=['Season'], keep='first') #Get rid of duplicates
highest_avg_score_season #Show team with highest average winning score for each season

Season  Wteam      Wscore
174     1985   1328   92.800000
287     1986   1109   91.200000
783     1987   1380   95.875000
977     1988   1258  111.750000
1268    1989   1258  117.315789
1560    1990   1258  126.347826
1950    1991   1380  112.312500
2246    1992   1380   99.642857
2542    1993   1380  101.875000
2842    1994   1380  106.583333
3003    1995   1206  102.833333
3305    1996   1206  103.600000
3648    1997   1254   94.142857
4070    1998   1395  101.153846
4312    1999   1317   95.750000
4695    2000   1395   95.647059
5013    2001   1395   99.315789
5201    2002   1242   92.888889
5660    2003   1395   88.222222
5809    2004   1194   92.142857
6239    2005   1311   91.333333
6460    2006   1190   89.533333
7026    2007   1440   99.400000
7367    2008   1440   94.500000
7652    2009   1377   96.200000
8061    2010   1440   95.000000
8414    2011   1449   90.130435
8683    2012   1370   87.400000
9036    2013   1377   86.500000
9333    2014   1322   92.600000
9683    2015   1322   90.733333
9861    2016   1146   97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [98]:
# Write your code here
win_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='NumWins') #Grouped by 'Season' and 'Wteam' and number of wins
win_count #show win count for each team in each season

Season  Wteam  NumWins
0        1985   1102        5
1        1985   1103        9
2        1985   1104       21
3        1985   1106       10
4        1985   1108       19
...       ...    ...      ...
10167    2016   1460       20
10168    2016   1461       12
10169    2016   1462       27
10170    2016   1463       21
10171    2016   1464        9

[10172 rows x 3 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [99]:
# Write your code here
team_wins_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='NumWins') #Similar as last task
most_win_index = team_wins_season.groupby('Season')['NumWins'].idxmax() #Find index of most wins team for each season
most_win_team_seasons = team_wins_season.loc[most_win_index] #Used index to find team with most wins each season
most_win_team_seasons #Show the team with most wins each season

Season  Wteam  NumWins
217     1985   1385       27
342     1986   1181       32
818     1987   1424       33
861     1988   1112       31
1323    1989   1328       28
1551    1990   1247       29
1739    1991   1116       30
2088    1992   1181       28
2423    1993   1231       28
2665    1994   1163       27
2928    1995   1116       26
3357    1996   1269       31
3638    1997   1242       31
3943    1998   1242       33
4201    1999   1181       32
4709    2000   1409       29
4829    2001   1181       29
5121    2002   1153       30
5456    2003   1166       29
5981    2004   1390       29
6165    2005   1228       32
6452    2006   1181       30
6841    2007   1242       30
7209    2008   1272       33
7552    2009   1272       31
7870    2010   1242       32
8217    2011   1242       32
8565    2012   1246       32
8878    2013   1211       30
9462    2014   1455       33
9609    2015   1246       34
9955    2016   1242       29

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [102]:
# Write your code here
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean() #Calculate losing score
average_winning_score_by_team = average_scores #Use win score from task 3

#Convert to dataframes
average_losing_score_df = average_losing_score_by_team.reset_index()
average_winning_score_df = average_winning_score_by_team.reset_index()
average_scores_comparison = pd.merge(average_winning_score_df, average_losing_score_df, left_on='Wteam', right_on='Lteam', how='outer') #Merge df

#Compare avg scores
average_scores_comparison['HigherAvgLosingScore'] = average_scores_comparison['Lscore'] > average_scores_comparison['Wscore']
teams_with_higher_avg_losing_scores = average_scores_comparison[average_scores_comparison['HigherAvgLosingScore']]

#Show teams with a higher average losing score than winning score
teams_with_higher_avg_losing_scores[['Wteam', 'Wscore', 'Lscore']]

Wteam     Wscore     Lscore
92      1104  63.500000  64.374317
116     1106  62.133333  66.284165
118     1106  65.444444  66.284165
123     1106  64.200000  66.284165
148     1108  62.166667  65.429825
...      ...        ...        ...
9829    1392  58.600000  59.164062
9946    1394  58.000000  61.450000
10093   1342  57.000000  58.051813
10112   1136  53.000000  59.802469
10141   1370  65.857143  66.559055

[211 rows x 3 columns]

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [71]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season    1985
Daynum      20
Wteam     1228
Wscore      81
Lteam     1328
Lscore      64
Wloc         N
Numot        0
Name: 0, dtype: object
Season    1985
Daynum      25
Wteam     1106
Wscore      77
Lteam     1354
Lscore      70
Wloc         H
Numot        0
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [104]:
# Write your code here
# Define a function to apply to each row
def high_scoring_game(row):
    if row['Wscore'] > 100:
        return 'Yes'
    else:
        return 'No'

# Create the new column by applying the function to each row
df['HighScoringGame'] = df.apply(high_scoring_game, axis=1) #Used apply instead of iterrows as it is more efficient in this case
df #Show table with HighScoringGame

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

       HighScoringGame  ScoreDifference  
0                   No               17  
1                   No                7  
2                   No                7  
3                   No               16  
4                   No               12  
...                ...              ...  
145284              No               20  
145285              No               14  
145286              No                5  
145287              No                4  
145288              No               13  

[145289 rows x 10 columns]

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [106]:
# Write your code here
# Initialize dictionary game_played to store games played
games_played = {}

# Iterate over each row in the DataFrame
for index, row in df.iterrows():
    # Increment the games played count for the winning team
    if row['Wteam'] in games_played:
        games_played[row['Wteam']] += 1
    else:
        games_played[row['Wteam']] = 1

    # Increment the games played count for the losing team
    if row['Lteam'] in games_played:
        games_played[row['Lteam']] += 1
    else:
        games_played[row['Lteam']] = 1

games_played #Show dictionary containing total number of games played by each team

{1228: 992,
 1328: 968,
 1106: 855,
 1354: 906,
 1112: 981,
 1223: 363,
 1165: 833,
 1432: 69,
 1192: 908,
 1447: 903,
 1218: 931,
 1337: 922,
 1226: 847,
 1242: 993,
 1268: 969,
 1260: 914,
 1133: 949,
 1305: 922,
 1424: 974,
 1307: 969,
 1288: 925,
 1344: 951,
 1438: 952,
 1374: 916,
 1411: 903,
 1412: 962,
 1397: 963,
 1417: 966,
 1225: 880,
 1116: 980,
 1368: 808,
 1120: 936,
 1391: 879,
 1135: 847,
 1306: 898,
 1143: 947,
 1388: 897,
 1153: 970,
 1184: 863,
 1159: 887,
 1171: 829,
 1216: 930,
 1173: 960,
 1134: 200,
 1177: 942,
 1296: 879,
 1193: 942,
 1265: 934,
 1196: 981,
 1416: 881,
 1206: 938,
 1137: 912,
 1210: 972,
 1149: 824,
 1211: 921,
 1102: 840,
 1234: 968,
 1114: 910,
 1332: 927,
 1243: 927,
 1317: 883,
 1257: 994,
 1231: 969,
 1277: 966,
 1145: 934,
 1278: 948,
 1453: 912,
 1286: 851,
 1186: 849,
 1301: 979,
 1144: 850,
 1325: 942,
 1384: 887,
 1326: 968,
 1248: 896,
 1287: 857,
 1339: 879,
 1334: 899,
 1365: 907,
 1375: 896,
 1126: 906,
 1403: 939,
 1152: 865,
 1423

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [107]:
# Write your code here
# Initialize dictionary to store the highest score difference per season
highest_diff_seasons = {}

# Iterate over each row in the DataFrame
for index, row in df.iterrows():
    # Calculate the score difference for the current game
    current_diff = row['Wscore'] - row['Lscore']

    # If the current season is not in the dictionary or the current difference is higher than the recorded one, update it
    if row['Season'] not in highest_diff_seasons or highest_diff_seasons[row['Season']]['score_diff'] < current_diff:
        highest_diff_seasons[row['Season']] = {'game_id': index, 'score_diff': current_diff, 'Wteam': row['Wteam'], 'Lteam': row['Lteam']}

highest_diff_seasons #Show dictionary containing the game with highest difference each season

{1985: {'game_id': 236, 'score_diff': 60, 'Wteam': 1361, 'Lteam': 1288},
 1986: {'game_id': 4731, 'score_diff': 84, 'Wteam': 1314, 'Lteam': 1264},
 1987: {'game_id': 8240, 'score_diff': 73, 'Wteam': 1155, 'Lteam': 1118},
 1988: {'game_id': 12046, 'score_diff': 68, 'Wteam': 1328, 'Lteam': 1147},
 1989: {'game_id': 16677, 'score_diff': 70, 'Wteam': 1242, 'Lteam': 1135},
 1990: {'game_id': 19502, 'score_diff': 76, 'Wteam': 1181, 'Lteam': 1217},
 1991: {'game_id': 25161, 'score_diff': 68, 'Wteam': 1163, 'Lteam': 1148},
 1992: {'game_id': 27997, 'score_diff': 82, 'Wteam': 1116, 'Lteam': 1126},
 1993: {'game_id': 33858, 'score_diff': 81, 'Wteam': 1328, 'Lteam': 1197},
 1994: {'game_id': 36404, 'score_diff': 69, 'Wteam': 1228, 'Lteam': 1152},
 1995: {'game_id': 39858, 'score_diff': 74, 'Wteam': 1246, 'Lteam': 1404},
 1996: {'game_id': 44653, 'score_diff': 91, 'Wteam': 1409, 'Lteam': 1341},
 1997: {'game_id': 49033, 'score_diff': 80, 'Wteam': 1278, 'Lteam': 1106},
 1998: {'game_id': 52600, 'sc

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [75]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [109]:
# Write your code here
# Count number of wins for each team
wins_count = df['Wteam'].value_counts() #Using value_counts()

# Count number of losses for each team
losses_count = df['Lteam'].value_counts() #Using value_counts()

# Add both Series together to get total number of games played by each team
total_games = wins_count.add(losses_count, fill_value=0)

total_games #Show total number of games

1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Length: 364, dtype: int64

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [110]:
# Write your code here
#Create new column for score difference
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

#Grouped by season and found index of the game with highest score difference
idx = df.groupby('Season')['ScoreDifference'].idxmax()

#Used index to select rows that contain the game with the highest score difference each season
games_with_highest_score_diff = df.loc[idx]

games_with_highest_score_diff #Show games with highest score diff each season

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410      87    H      0   
57618     1999      58   1268     132   1317      57    H      0   
60616     2000      19   1261     112   1212      37    H      0   
65999     2001      41   1196     125   1197      50    H      0   
70737     2002      50   1328     109   1183      50    H      0   
74791     2003      33   1403     107   1311      35    H      0   
79181     2004      29   1328      94   1115      24    H      0   
83372     2005      18   1403     119   1421      55    H      0   
89542     2006      60   1400     110   1341      38    H      0   
94507     2007      59   1401     101   1212      27    H      0   
97737     2008       4   1182     129   1224      59    H      0   
103723    2009      27   1314     116   1421      48    H      0   
108405    2010      15   1397     124   1421      49    H      0   
114015    2011      22   1388     121   1152      52    H      0   
118699    2012      11   1345      96   1296      34    H      0   
124057    2013       7   1235      98   1105      40    H      0   
131373    2014      61   1256     126   1255      52    H      0   
134590    2015      11   1181     113   1342      44    H      0   
140040    2016      12   1139     144   1154      71    H      0   

       HighScoringGame  ScoreDifference  
236                Yes               60  
4731               Yes               84  
8240               Yes               73  
12046              Yes               68  
16677              Yes               70  
19502              Yes               76  
25161              Yes               68  
27997              Yes               82  
33858              Yes               81  
36404              Yes               69  
39858              Yes               74  
44653              Yes               91  
49033              Yes               80  
52600              Yes               66  
57618              Yes               75  
60616              Yes               75  
65999              Yes               75  
70737              Yes               59  
74791              Yes               72  
79181               No               70  
83372              Yes               64  
89542              Yes               72  
94507              Yes               74  
97737              Yes               70  
103723             Yes               68  
108405             Yes               75  
114015             Yes               69  
118699              No               62  
124057              No               58  
131373             Yes               74  
134590             Yes               69  
140040             Yes               73

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [38]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [39]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [40]:
type(df['Wscore'])

pandas.core.series.Series

In [41]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [42]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

Here's an equivalent using iloc

In [43]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [44]:
df.isnull().sum()

Season    0
Daynum    0
Wteam     0
Wscore    0
Lteam     0
Lscore    0
Wloc      0
Numot     0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y